python/src/robyn/modeling/ridge_model_builder.py

```diff
@@ -582,80 +531,90 @@ def _evaluate_model(
             y_val_pred = model.predict(X_val)
             y_test_pred = model.predict(X_test)
-            nrmse_val = np.sqrt(np.mean((y_val - y_val_pred) ** 2)) / (y_val.max() - y_val.min())
-            nrmse_test = np.sqrt(np.mean((y_test - y_test_pred) ** 2)) / (y_test.max() - y_test.min())
             rsq_val = r2_score(y_val, y_val_pred)
             rsq_test = r2_score(y_test, y_test_pred)
-            nrmse = nrmse_val  # Use validation NRMSE for optimization
+            nrmse_val = np.sqrt(mean_squared_error(y_val, y_val_pred)) / (y_val.max() - y_val.min())
+            nrmse_test = np.sqrt(mean_squared_error(y_test, y_test_pred)) / (y_test.max() - y_test.min())
+            nrmse = nrmse_val
         else:
-            y_val_pred = y_test_pred = None
-            nrmse_val = nrmse_test = None
+            # Set default values for non-validation metrics
             rsq_val = rsq_test = 0.0
-            nrmse = nrmse_train  # Use training NRMSE when not doing validation
+            nrmse_val = nrmse_test = 0.0
+            nrmse = nrmse_train
 
+        # Calculate RSSD and MAPE
         decomp_rssd = self._calculate_rssd(model.coef_, rssd_zero_penalty)
-        mape = self._calculate_mape(model) if self.calibration_input else 0.0
-        lift_calibration = self._calculate_lift_calibration(model) if self.calibration_input else None
-
-        lambda_hp = params.get("lambda", 0.0)
-        lambda_max = self._lambda_seq(X_train, y_train)[0]
-        lambda_min_ratio = 0.0001
 
-        pos = np.all(model.coef_ >= 0)
-
-        if objective_weights is None:
-            objective_weights = [1 / 3, 1 / 3, 1 / 3] if self.calibration_input else [0.5, 0.5]
-
-        loss = (
-            objective_weights[0] * nrmse
-            + objective_weights[1] * decomp_rssd
-            + (objective_weights[2] * mape if self.calibration_input else 0)
-        )
-
-        elapsed = time.time() - start_time
+        # Handle calibration
+        lift_calibration = None
+        if self.calibration_input is not None:
+            if hasattr(self, "calibration_engine"):
+                calibration_engine = self.calibration_engine
+            else:
+                calibration_engine = MediaEffectCalibrator(
+                    mmm_data=self.mmm_data,
+                    hyperparameters=self.hyperparameters,
+                    calibration_input=self.calibration_input,
+                )
+                self.calibration_engine = calibration_engine
+
+            lift_collect = calibration_engine.calibrate()
+            if lift_collect is not None:
+                lift_calibration = lift_collect
+                mape = float(lift_collect.get_mean_mape())
+                calibration_constraint = 0.1
+                if mape > calibration_constraint:
+                    decomp_rssd *= 1 + (mape - calibration_constraint)
+            else:
+                mape = 0.0
+        else:
+            mape = 0.0
```

Comment:
nit; just add mape = 0.0 before if block so you don't need two else blocks above.